### PR TITLE
BRIN changes to user interface

### PR DESCRIPTION
**Pull request by soumyadeep2007**
Please review commits individually. Mainly consists of 2 changes:
1. Making brin_desummarize_range MPP aware.
2. Warning users when the autosummarize reloption is set for BRIN indexes.

Dev pipeline: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/brin_interface

### CODE REVIEW

**Comment by soumyadeep2007** [comment]
Fixed optimizer file for 1st commit in latest force push.

File: src/test/regress/expected/brin_interface.out [hunk 1]
```diff
@@ -144,5 +144,15 @@ SELECT gp_execution_segment(), (brin_page_items(get_raw_page('brin_interface_i_i
...
+-- Don't warn users when autosummarize is specified as false during index build.
+CREATE INDEX ON brin_interface USING brin(i) WITH (autosummarize=false);
+-- Warn users when autosummarize is specified as true during index build.
+CREATE INDEX brin_interface_warn_idx ON brin_interface USING brin(i) WITH (autosummarize=true);
+WARNING:  autosummarize is not supported, setting it will have no effect
+-- Altering it to false should not result in a warning.
+ALTER INDEX brin_interface_warn_idx SET (autosummarize=false);
+-- Altering it to true should result in a warning.
+ALTER INDEX brin_interface_warn_idx SET (autosummarize=true);
+WARNING:  autosummarize is not supported, setting it will have no effect
```
Comment (huansong) [inline src/test/regress/expected/brin_interface.out:156]: nit: the `.sql` changes are in the other commit. Squash would be fine.

File: src/backend/access/brin/brin.c [hunk 2]
```diff
@@ -1061,6 +1062,15 @@ brinoptions(Datum reloptions, bool validate)
...
 				   validate, tab, lengthof(tab));
 
+	/*
+	 * GPDB: We don't support autosummarize yet, as we don't support user-table
+	 * autovacuum. So, WARN users accordingly.
+	 */
+	if (IS_QUERY_DISPATCHER() && validate && rdopts->autosummarize)
+		ereport(WARNING,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("autosummarize is not supported, setting it will have no effect")));
```
Comment (huansong) [inline src/backend/access/brin/brin.c:1072]: Q about autosummarize: the [doc](https://www.postgresql.org/docs/current/brin-intro.html#:~:text=regardless%20of%20whether%20the%20table%20itself%20is%20processed%20by%20autovacuum) says:
> if the index's autosummarize parameter is enabled, which it isn't by default, whenever autovacuum runs in that database, summarization will occur for all unsummarized page ranges that have been filled, regardless of **whether the table itself is processed by autovacuum**

From the code it also looks like we are just sending work to the autovacuum process through `AutoVacuumRequestWork`. Doesn't that mean even we don't support user table autovacuum, we might still do autosummarize?

File: src/test/regress/sql/brin_interface.sql [hunk 3]
```diff
@@ -0,0 +1,54 @@
+-- Additional tests that exercise the BRIN user interface
+CREATE EXTENSION pageinspect;
+
+CREATE TABLE brin_interface(i int);
+INSERT INTO brin_interface SELECT * FROM generate_series(1, 5000);
+CREATE INDEX ON brin_interface USING brin(i) WITH (pages_per_range=1);
```
Comment (huansong) [inline src/test/regress/sql/brin_interface.sql:6]: General Q: wondering about the purpose of this test. Seems that with this PR we are not making changes to the interface (except for the return type of `brin_desummarize_range`). Would we have other interface change in future (that is not covered by the existing `regress/brin` and `isolation2/brin` tests)? But in any case I'm all for additional coverage.

**Review by huansong — COMMENTED** [review]
Just a Q about autosummarize. The rest LGTM.

File: src/test/regress/expected/brin_interface.out (hunk 1 above)
Comment (soumyadeep2007) [reply src/test/regress/expected/brin_interface.out:156]: Ah sorry let me actually separate those.

File: src/test/regress/sql/brin_interface.sql (hunk 3 above)
Comment (soumyadeep2007) [reply src/test/regress/sql/brin_interface.sql:6]: The other tests are on data co-located on 1 segment for simplicity. So this is just adding additional coverage to test MPPness of all brin funcs.

File: src/backend/access/brin/brin.c (hunk 2 above)
Comment (soumyadeep2007) [reply src/backend/access/brin/brin.c:1072]: Ah you are very right. I see that with this interface we completely bypass the check for `relation_needs_vacanalyze`, which checks for catalog namespace etc.

I'll change this WARNING to an ERROR.

**Review by huansong — APPROVED** [review]
LGTM!

File: src/test/regress/sql/brin_interface.sql (hunk 3 above)
Comment (huansong) [reply src/test/regress/sql/brin_interface.sql:6]: Got it, thanks for clarification!

**Pull request merged**